dashboard/src/components/AppUpgrade/AppUpgrade.tsx

```diff
@@ -12,7 +12,8 @@ export interface IAppUpgradeProps {
   app?: IRelease;
   appsIsFetching: boolean;
   chartsIsFetching: boolean;
-  appsError: Error | undefined;
+  getError?: Error;
+  upgradeError?: Error;
```

Comment:
It's not clear to me why you're adding different error props (get, upgrade) with a plain Error type, rather than just leaving the one error prop and [custom error types](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types)? Wouldn't that be neater (and not lead to adding an error prop per error type across components)?

Reply:
I am not sure if that would be more neat? For two reasons:
 - It's technically possible (although it won't happen) that there were two errors colliding (error fetching an deleting an app for example). If there were only one prop for the error, we would lose information about one of them.
 - We would need to check the error type anyway. For example, in this PR, we would need to check `if (err.Constructor === FetchError)` or alike in the `AppUpgrade` component, and in case it's not, forward the error to the `UpgradeForm` which is not much more neat.

Are you thinking on something I'm missing?

Reply:
> I am not sure if that would be more neat?

Neater that the alternative of adding a new error prop for each possible error? :P 

> For two reasons:
> 
>     * It's technically possible (although it won't happen) that there were two errors colliding (error fetching an deleting an app for example). If there were only one prop for the error, we would lose information about one of them.

I'm not sure what the issue is here - we normally don't continue after the first error - and even if this was possible in practise, it's not hard to create a [custom error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#ES6_Custom_Error_Class) handling whatever situation you want.

> 
>     * We would need to check the error type anyway. For example, in this PR, we would need to check `if (err.Constructor === FetchError)` or alike in the `AppUpgrade` component, and in case it's not, forward the error to the `UpgradeForm` which is not much more neat.

The neatness isn't in the handling of the error, but in not having to add an extra prop for each type of error we can encounter here.

That said, I think the handling can be just as neat. Here are some different scenarios. Let's say we were handling different types of errors in the one spot:

```
switch (err.Constructor) {
  case FetchError:
    ...
    break;
  case ...
}
```

Or just wanting to handle one type of error as you've mentioned:
```
if (err.Constructor === FetchError) {...}
```

although, I'm not saying you should, but remember you can put whatever syntactical sugar you like in our error handling, so it could look like:

```
if errors.IsA(err, FetchError) {}
```

I think your last point (correct me - not 100%) is that in the `AppUpgrade` component you'd need to check the error type and so conditionally forward it to the `UpgradeForm`? I'd just set the error prop unconditionally, you already handle a fetch error here and don't continue:
```
  if (errors.IsA(error, FetchError)) {
    return <Alert theme="danger">Unable to retrieve the current app: {error.message}</Alert>;
  }
  ...
    <UpgradeForm
    ...
    error={error}
    ...
    />
```

So I'd expect the `UpgradeForm` to do the same:
```
if (errors.IsA(error, UpgradeError)) { 
    return <Alert theme="danger">Some specific upgrade error handling</Alert>;
}
```

(Unrelated to this discussion, you might even want to throw the error when a leaf component doesn't handle an error so it's caught and handled generically by the error boundary if our specific handling doesn't handle it).

> 
> 
> Are you thinking on something I'm missing?

Only that the issue of neatness is, in my mind, centered around whether we need to add an extra prop for each different type of error, as opposed to handling a single error according to its type.

I've +1'd earlier, I just worry that this creeps in as a pattern in other places (or already has) so would be great to think about and handle the right way now. Up to you.

Reply:
> > I am not sure if that would be more neat?
> 
> Neater that the alternative of adding a new error prop for each possible error? :P
> 

You would need to add a new error class for each possible error anyway (that's why I say it's more or less the same).

> > For two reasons:
> > ```
> > * It's technically possible (although it won't happen) that there were two errors colliding (error fetching an deleting an app for example). If there were only one prop for the error, we would lose information about one of them.
> > ```
> 
> I'm not sure what the issue is here - we normally don't continue after the first error - and even if this was possible in practise, it's not hard to create a [custom error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#ES6_Custom_Error_Class) handling whatever situation you want.
> 

No, the thing is that if we only store one error, we could overwrite the original error. I am thinking about this case:

 - A request is send to delete app A
 - The deletion is done but some error occurred so we set a deletion error
 - After the deletion is done, a refresh is trigger to retrieve information again
 - An error occurs because A no longer exists so we set a fetch error

The fetch error above would remove the information about the deletion error (which is the important one). That scenario doesn't happen IRL but that was what I meant.

> I think your last point (correct me - not 100%) is that in the `AppUpgrade` component you'd need to check the error type and so conditionally forward it to the `UpgradeForm`? I'd just set the error prop unconditionally, you already handle a fetch error here and don't continue:
> 
> ```
>   if (errors.IsA(error, FetchError)) {
>     return <Alert theme="danger">Unable to retrieve the current app: {error.message}</Alert>;
>   }
>   ...
>     <UpgradeForm
>     ...
>     error={error}
>     ...
>     />
> ```
> 
> So I'd expect the `UpgradeForm` to do the same:
> 
> ```
> if (errors.IsA(error, UpgradeError)) { 
>     return <Alert theme="danger">Some specific upgrade error handling</Alert>;
> }
> ```

That was the behavior prior to this PR, the goal is to not return when there is an upgrade error to allow people to modify the UpgradeForm with new data so the error can be fixed, without losing already introduced input.

> (Unrelated to this discussion, you might even want to throw the error when a leaf component doesn't handle an error so it's caught and handled generically by the error boundary if our specific handling doesn't handle it).

We do that with the error boundary, but that is only for unexpected errors thrown (not errors set in the Redux state). We could change the approach and have all the errors handled by the error boundary but that would require a fairly big refactor.

> > Are you thinking on something I'm missing?
> 
> Only that the issue of neatness is, in my mind, centered around whether we need to add an extra prop for each different type of error, as opposed to handling a single error according to its type.
> 
> I've +1'd earlier, I just worry that this creeps in as a pattern in other places (or already has) so would be great to think about and handle the right way now. Up to you.

You still need to add error classes for each different type of error but it's true that it can be reused in other scenarios (so `FetchError` can be used both for apps and operators for example).